### PR TITLE
JAMES-2107 Java-8 migration + automatic refactoring

### PR DESCRIPTION
**Pull request by chibenwa**
Includes automatic IntelliJ refactorings:
 - From imperative to Streams
 - From anonymous to lambdas
 - Remove no more needed explicit
 - Some call chain simplification

This PR is based on https://github.com/linagora/james-project/pull/918

It is run directly by a custom CI JOB as it is not compatible with CI current state. See https://github.com/linagora/james-jenkins/pull/77. 

Direct link to CI Job: https://james.open-paas.org/job/build%20JAMES-2107-2/

### CODE REVIEW

**Comment by chibenwa** [comment]
Build is green here

File: mailbox/api/src/main/java/org/apache/james/mailbox/acl/SimpleGroupMembershipResolver.java [hunk 1]
```diff
@@ -68,7 +68,7 @@ public String toString() {
 
     }
 
-    private final Set<Membership> memberships = new HashSet<SimpleGroupMembershipResolver.Membership>(32);
+    private final Set<Membership> memberships = new HashSet<>(32);
```
Comment (quynhn) [inline mailbox/api/src/main/java/org/apache/james/mailbox/acl/SimpleGroupMembershipResolver.java]: It's potential have an eclipse warning

Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/acl/SimpleGroupMembershipResolver.java]: Let's eclipse guys correct this later on where need be?

It is not in my IDE, and no ERRROR raised in **maven**. However Eclipse guys raise a lot of errors in my IDE :trollface:  .

File: README.adoc [hunk 2]
```diff
@@ -115,17 +92,22 @@ First step, you have to build the Docker image
 
 In order to run the build, you have to launch the following command:
 
-    $ docker run -v $PWD/.m2:/root/.m2 -v $PWD:/origin -v $PWD/dockerfiles/run/spring/destination:/destination -t james/project -s SHA1
+    $ docker run -v $PWD/.m2:/root/.m2 -v $PWD:/origin -t james/project -s SHA1
```
Comment (mbaechler) [inline README.adoc]: I liked the previous version more : it doesn't involve reading things below

Comment (mbaechler) [inline README.adoc]: I think we could simply do a multi-stage build : https://docs.docker.com/engine/userguide/eng-image/multistage-build/

File: src/site/xdoc/server/install.xml [hunk 3]
```diff
@@ -29,14 +29,11 @@
 
   <subsection name="Java Virtual Machine">
   
-    <p>James requires a Java Runtime Environment. Java version 1.6 or higher is required to run the 
+    <p>James requires a Java Runtime Environment. Java version 1.6 or higher is required to run the
```
Comment (mbaechler) [inline src/site/xdoc/server/install.xml]: 1.8

File: backends-common/pom.xml [hunk 4]
```diff
@@ -68,6 +68,11 @@
                 <version>${junit.version}</version>
             </dependency>
             <dependency>
+                <groupId>org.assertj</groupId>
+                <artifactId>assertj-core</artifactId>
+                <version>${assertj-3.version}</version>
```
Comment (mbaechler) [inline backends-common/pom.xml]: the property should not include 3 anymore

File: backends-common/jpa/src/test/java/org/apache/james/backends/jpa/JpaTestCluster.java [hunk 5]
```diff
@@ -58,12 +58,7 @@ public static JpaTestCluster create(List<Class<?>> clazz) {
     }
 
     private static Function<Class<?>, String> toFQDN() {
```
Comment (mbaechler) [inline backends-common/jpa/src/test/java/org/apache/james/backends/jpa/JpaTestCluster.java]: it looks like you can inline it

File: mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerStressTest.java [hunk 6]
```diff
@@ -93,37 +93,34 @@ public void event(Event event) {
...
+                if (fail.get()) {
+                    latch.countDown();
+                    return;
+                }
 
 
-                    try {
-                        MailboxSession session = mailboxManager.createSystemSession(username, LoggerFactory.getLogger("Test"));
+                try {
+                    MailboxSession session1 = mailboxManager.createSystemSession(username, LoggerFactory.getLogger("Test"));
```
Comment (mbaechler) [inline mailbox/api/src/test/java/org/apache/james/mailbox/MailboxManagerStressTest.java]: refactoring failure : the variable name is wrongly renamed

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java [hunk 7]
```diff
@@ -66,12 +66,7 @@
 public class JPAMessageMapper extends JPATransactionalMapper implements MessageMapper {
     private static final int UNLIMIT_MAX_SIZE = -1;
     private static final int UNLIMITED = -1;
-    private static final Function<MailboxMessage, MessageUid> TO_UID = new Function<MailboxMessage, MessageUid>() {
-        @Override
-        public MessageUid apply(MailboxMessage mailboxMessage) {
-            return mailboxMessage.getUid();
-        }
-    };
+    private static final Function<MailboxMessage, MessageUid> TO_UID = MailboxMessage::getUid;
```
Comment (mbaechler) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: should be inline

**Comment by mbaechler** [comment]
Just as a friendly advice : you could split the PR and have it merged by chunks to avoid struggling with a long lasting PR.

File: README.adoc (hunk 2 above)
Comment (chibenwa) [inline README.adoc]:
```
Multi-stage builds are a new feature in Docker 17.05
```

In my dev environment:

```
benwa@horizon ~/Documents/apache/onssd/james-project (git)-[JAMES-2048] % docker -v
Docker version 17.06.0-ce, build 3dfb8343
```

But on Yoda:

```
root@yoda:/home/btellier# docker -v
Docker version 1.11.2, build b9f10c9
```

Which means Docker upgrade on online servers....

Moreover, you are proposing a fancy new tech in a review about getting the job done. Maybe not the right place for this. I believe we should uncouple Java-8 migration from multi-stage builds. After all maybe this shiny proposal deserve a ticket by it's own, no?

File: src/site/xdoc/server/install.xml (hunk 3 above)
Comment (chibenwa) [inline src/site/xdoc/server/install.xml]: s/James/James 3.0/g that is the true issue here.

File: README.adoc (hunk 2 above)
Comment (chibenwa) [inline README.adoc]: The previous build instruction is partial and is IMO misleading. I do know only about one specific outpout. Here I know about several. I still defend it.

**Comment by chibenwa** [comment]
https://github.com/linagora/james-project/pull/918 is the part of this PR without automatic refactorings.

@mbaechler Do you expect me to split it more?

File: README.adoc (hunk 2 above)
Comment (mbaechler) [inline README.adoc]: I agree about the problem of adopting a new docker feature. Nonetheless, we should propose a strategy that doesn't involve copy/pasting a command line then understand that I should have mounted a volume and do it again.

**Comment by mbaechler** [comment]
nope, should be ok

**Comment by chibenwa** [comment]
Green on CI

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**